File: aws-cdk/src/main/java/com/rudolfs/gatling/cdk/vpc/GatlingVpcStack.java
package com.rudolfs.gatling.cdk.vpc;

import software.amazon.awscdk.core.Construct;
import software.amazon.awscdk.core.Stack;
import software.amazon.awscdk.core.StackProps;
import software.amazon.awscdk.core.Tag;
import software.amazon.awscdk.core.TagProps;
import software.amazon.awscdk.services.ec2.IVpc;
import software.amazon.awscdk.services.ec2.SubnetConfiguration;
import software.amazon.awscdk.services.ec2.SubnetType;
import software.amazon.awscdk.services.ec2.Vpc;

import java.util.List;
import java.util.function.Supplier;

import static java.util.Collections.singletonList;

/**
 * Creates the CloudFormation for an AWS VPC with a private subnet and public subnet (both redundant in two availability zones).
 */
public class GatlingVpcStack extends Stack implements Supplier<IVpc> {
    private final Vpc vpc;

    public GatlingVpcStack(Construct scope, String id, StackProps stackProps, String namespace) {
        super(scope, id, stackProps);

        SubnetConfiguration privateSubnet = subnet("gatling-private-subnet", 19, SubnetType.PRIVATE);
        SubnetConfiguration publicSubnet = subnet("gatling-public-subnet", 20, SubnetType.PUBLIC);

        this.vpc = Vpc.Builder.create(this, "GatlingVpc")
                .cidr("10.12.0.0/16")
                .maxAzs(2)
                .subnetConfiguration(List.of(privateSubnet, publicSubnet))
                .build();

        Tag.add(vpc, "Name", namespace + "-vpc", TagProps.builder()
                .includeResourceTypes(singletonList("AWS::EC2::VPC"))
                .build());
    }

    SubnetConfiguration subnet(String name, Number cidrMask, SubnetType subnetType) {
        return SubnetConfiguration.builder()
                .cidrMask(cidrMask)
                .name(name)
                .subnetType(subnetType)
                .build();
    }

    @Override
    public IVpc get() {
        return this.vpc;
    }
}

File: aws-cdk/aws-cdk-typescript/lib/gatling-vpc-stack.ts
import * as cdk from '@aws-cdk/core';
import { Vpc, SubnetConfiguration, SubnetType, IVpc } from '@aws-cdk/aws-ec2';

export class GatlingVpcStack extends cdk.Stack {
  public vpc: Vpc;

  constructor(scope: cdk.Construct, id: string, namespace: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const privateSubnet: SubnetConfiguration = {
      name: 'gatling-private-subnet',
      cidrMask: 19,
      subnetType: SubnetType.PRIVATE
    };

    const publicSubnet: SubnetConfiguration = {
      name: 'gatling-public-subnet',
      cidrMask: 20,
      subnetType: SubnetType.PUBLIC
    };

    this.vpc = new Vpc(this, 'GatlingVpc', {
      cidr: '10.12.0.0/16',
      maxAzs: 2,
      subnetConfiguration: [privateSubnet, publicSubnet]
    });

    cdk.Tag.add(this.vpc, 'Name', `${namespace}-vpc`, {includeResourceTypes: ['AWS::EC2::VPC']} );
  }
}

export class ExistingVpcStack extends cdk.Stack {
  public vpc: IVpc;

	constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps, vpcName?: string) {
    super(scope, id, props);
  
    this.vpc = Vpc.fromLookup(this, 'gatlingVpc', {
      vpcName: vpcName
    });
  }
}
